juju/model.py

```diff
@@ -2078,7 +2078,7 @@ def _upload(self, data, path, app_name, res_name, res_type, pending_id):
 
         headers['Content-Type'] = 'application/octet-stream'
         headers['Content-Length'] = len(data)
-        headers['Content-Sha384'] = hashlib.sha384(bytes(data, 'utf-8')).hexdigest()
+        headers['Content-Sha384'] = hashlib.sha384(data).hexdigest()
```

Comment:
This effectively changes the argument type from `str` to `bytes`.

There's one more call site at line 2141 that would have to be updated to match.

P.S. please add an explicit type hint on the argument.

Reply:
done now